### PR TITLE
Navigation: Load the raw property on the navigation fallback

### PR DESCRIPTION
**Pull request by scruffian**
## What?
When loading the navigation fallback with an embed context we need to add the raw property of the title so that we have the response in the format expected by the frontend.

## Why?
Without the raw property, the data resolver treats the data differently which causes errors like https://github.com/WordPress/gutenberg/issues/52680 and https://github.com/WordPress/gutenberg/issues/52691

## How?
Updates the schema to include the raw property.

## Testing Instructions
1. Open the site editor at the root level
2. Open the patterns section
3. Open the headers section
4. Click on a header
5. Add a navigation block to the header, making sure you have two navigation blocks in the template
6. Make sure that the two navigation blocks use different wp_navigation CPTs
7. In trunk you should see an error, on this branch you should not.

## Note
I tried adding an end to end test for this change, but it's not possible to navigate directly to the root of the site editor in the test environment - not sure why.

### CODE REVIEW

File: lib/compat/wordpress-6.3/navigation-fallback.php
Comment (getdave) [inline lib/compat/wordpress-6.3/navigation-fallback.php]:
```suggestion
	// These aren't exposed by the posts controller by default, see:
	// https://github.com/WordPress/wordpress-develop/blob/5c3c6258e468c67ba00bbd13db29994f1a57a52a/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2425.
```

File: lib/compat/wordpress-6.3/navigation-fallback.php
Comment (getdave) [inline lib/compat/wordpress-6.3/navigation-fallback.php]:
```suggestion
	// Expose sub properties of title field.
	// These aren't exposed by the posts controller by default, see:
	// https://github.com/WordPress/wordpress-develop/blob/5c3c6258e468c67ba00bbd13db29994f1a57a52a/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L2401.
	$schema['properties']['title']['properties']['raw']['context'] = array_merge( $schema['properties']['title']['properties']['raw']['context'], array( 'embed' ) );
```

**Review by getdave — COMMENTED** [review]
The change itself is solid. The precedent is set by the same changes to the `content` field which already got approval from REST API maintainers in WP Core.

I would add a PHPUnit test that asserts that making a call to the endpoint returns both `rendered` and `raw` fields.

We already have the tests so it's just a case of adding one more. These should assert on the presence of fields in the response that aren't typically included in `_embed` responses:

- content
- title.raw

We could even use the `test_context_param` test which is currently disabled.

Once done let's spin up that Core patch asap.

**Comment by t-hamano** [comment]
I have tested this PR and can confirm that the issue reported in #52691 has been resolved.

**Comment by getdave** [comment]
For the test I suggest that because [we can't work with `embedded` results in Test setup](https://core.trac.wordpress.org/ticket/55961), we instead just call the linked endpoint directly in the test with `context` set to `embed`  to simulate the exact same thing.

Something like this:

```php
public function test_adds_correct_fields() {
	$request  = new WP_REST_Request( 'GET', '/wp-block-editor/v1/navigation-fallback' );
	$response = rest_get_server()->dispatch( $request );
	$data     = $response->get_data();

	$navigation_post_id = $data['id'];

	$links = $response->get_links();
        
    // Manually request embedded data due to this issue - https://core.trac.wordpress.org/ticket/55961.
	$embeddable_post_rest_url = str_replace( 'http://localhost:8889/index.php?rest_route=', '', $links['self'][0]['href'] );

	$request  = new WP_REST_Request( 'GET', $embeddable_post_rest_url );
	$request->set_param( 'context', 'embed' );
	$response = rest_get_server()->dispatch( $request );
	$data     = $response->get_data();

    // Assertions here to test for correct fields being included.
	echo '<pre>';
	var_dump( $data );
	echo '</pre>';
}
```

_Ideally_ we’d not hard code this part http://localhost:8889/index.php?rest_route=. Ideally that would be dynamically created.

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$embedded_navigation_href = $links['self'][0]['href'];
		preg_match( '/\?rest_route=(.*)/', $embedded_navigation_href, $matches );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		// Fetch the "linked" navigation post from the endpoint, with the context parameter set to 'embed' to simulate fetching embedded links.
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$this->assertArrayHasKey( 'status', $data, 'Response title should contain a "status" field.' );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$this->assertArrayHasKey( 'content', $data, 'Response should contain a "content" field.' );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$this->assertArrayHasKey( 'raw', $data['content'], 'Response content should contain a "raw" field.' );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$this->assertArrayHasKey( 'rendered', $data['content'], 'Response content should contain a "rendered" field.' );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
		$this->assertArrayHasKey( 'block_version', $data['content'], 'Response should contain a "block_version" field.' );
```

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]: Nit: can we collocate the subfield checks for fields of the same type (e.g. content, title...etc)?

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
	public function test_embedded_navigation_post_contains_required_fields() {
```

Let's make this explicit.

File: phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php
Comment (getdave) [inline phpunit/class-gutenberg-rest-navigation-fallback-controller-test.php]:
```suggestion
	 * By default, the REST response for the Posts Controller will not return all fields
	 * when the context is set to 'embed'. Assert that correct additional fields are added
	 * to the embedded Navigation Post, when the navigation fallback endpoint 
	 * is called with the `_embed` param.
	 *
	 * @covers wp_add_fields_to_navigation_fallback_embeded_links
```

Nit...

**Review by getdave — COMMENTED** [review]
Looks good on the test. A few tweaks required 🙏

**Review by getdave — APPROVED** [review]
LGTM. I don't think test failures are relevant.

**Comment by getdave** [comment]
I've forced merged this after consulting other contributors and confirming these tests are failing for all PRs. They are unrelated to this PR and holding up an important bug fix that can crash the editor.

**Comment by ramonjd** [comment]
This PR needs a backport. Added label

**Comment by ramonjd** [comment]
RC2 for WordPress 6.3 will be cut in the next few hours. If folks don't have time, I can get one ready

**Comment by tellthemachines** [comment]
I just cherry-picked this PR to the update/packages-RC2 branch to get it included in the next release: c98607bc77

**Pull request merged**
